space-age/space_age.go

```diff
@@ -1,3 +1,39 @@
 package space
 
 type Planet string
+
+var (
+	earthSeconds float64 = 31557600
+)
+
+// Return earth-years old value for given seconds on specific planet
+func Age(seconds float64, planet Planet) float64 {
+	return seconds / planetOrbitalPeriod(planet)
+}
+
+func planetOrbitalPeriod(planet Planet) float64 {
+
+	// I wonder if there is a cleaner way of handling this.
+	// It was switch/case or if statements in my head.
+	// Both are a bit cluttered.
+	switch {
+	case planet == "Mercury":
+		return 0.2408467 * earthSeconds
+	case planet == "Venus":
+		return 0.61519726 * earthSeconds
+	case planet == "Earth":
+		return earthSeconds
+	case planet == "Mars":
+		return 1.8808158 * earthSeconds
+	case planet == "Jupiter":
+		return 11.862615 * earthSeconds
+	case planet == "Saturn":
+		return 29.447498 * earthSeconds
+	case planet == "Uranus":
+		return 84.016846 * earthSeconds
+	case planet == "Neptune":
+		return 164.79132 * earthSeconds
+	}
+
+	return earthSeconds
```

Comment:
This is a valid way of solving this, but probably not the most elegant.

By having a switch with lots of cases here you've effectively recreated a map, a map does what it says on the tin and maps one value to another.

Golang has very nice support for maps, and by their nature, looking up an item in a map is an `O(1)` time complexity (i.e. no matter how big the map gets it takes the same time to look things up). Look at this example and see how it could be applied above:

```golang
var jobs = map[string]string{
  "dan": "developer",
  "aaron": "devops",
}

func jobDescription(name string) string {
  return fmt.Sprintf("%s is a %s", name, jobs[name])
}
```

This example doesn't cover how to deal with an item that does not exist in the map, have a think about that too :D

Reply:
I think I've got the map part down now - but I'm not sure what'll happen if an item doesn't exist in the map.

Reply:
Yeah very good question. In this case you're a bit limited by the function signature that the test case expects (`Age(seconds float64, planet Planet)`). If you had the ability to return an error as well then you could return an error in the event that an element did not exist in the map.

There is a nice trick with Go maps, you can actually get two values from a map lookup. The first will be the value in the map, the second is a boolean showing you whether they is a value at the key you requested. As an example:

```golang
package main

import (
	"fmt"
	"log"
)

var jobs = map[string]string{
	"dan":   "developer",
	"aaron": "devops engineer",
}

func jobDescription(name string) (string, error) {
	job, ok := jobs[name]
	if !ok {
		return "", fmt.Errorf("no job for %q", name)
	}

	return fmt.Sprintf("%s is a %s", name, job), nil
}

func main() {
	if j, err := jobDescription("aaron"); err != nil {
		log.Fatal(err)
	} else {
		fmt.Println(j)
	}

	if j, err := jobDescription("bob"); err != nil {
		log.Fatal(err)
	} else {
		fmt.Println(j)
	}
}
```